### PR TITLE
2290 Add support to update data in Datasets

### LINKED ISSUE #2290: support dynamic data list for the dataset APIs (2/July)
**Is your feature request related to a problem? Please describe.**
At the workflow level, would be great to have the capability of changing the input training dataset on the fly, without restarting the running engine. This would require some extra works for the cache-based datasets.

cc @SachidanandAlle @diazandr3s

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #2290 .

### Description
This PR added support to update data after several epochs in Datasets.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/dataset.py [hunk 1]
```diff
@@ -173,6 +176,17 @@ def __init__(
             if not self.cache_dir.is_dir():
                 raise ValueError("cache_dir must be a directory.")
 
+    def update_data(self, data: Sequence):
```
Comment (wyli) [inline monai/data/dataset.py:179]: LMDBDataset inherits this but probably wouldn't work

**Review by wyli — COMMENTED** [review]
I think the new method could be `set_data` instead of `update_data`? not sure about the overall stability in the multiprocess loading because there will be concurrent read and write for the cache. maybe the PersistentDataset is ok but how about the other datasets?

File: monai/data/dataset.py [hunk 2]
```diff
@@ -173,6 +176,17 @@ def __init__(
...
+        """
+        Update the input data and delete all the out-dated cache content.
+
+        """
+        self.data = data
+
+        if self.cache_dir is not None and self.cache_dir.exists():
+            shutil.rmtree(self.cache_dir)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/data/dataset.py:187]: updated in the latest commit.
Thanks.

File: monai/data/dataset.py [hunk 3]
```diff
@@ -711,6 +744,22 @@ def __init__(
...
+
+        """
+        if self.is_started():
+            warnings.warn("SmartCacheDataset is not shutdown yet, shutdown it directly.")
+            self.shutdown()
+
+        if self.shuffle:
+            data = copy(data)
+            self.randomize(data)
+        return super().update_data(data)
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply monai/data/dataset.py:761]: Thanks for pointing out it, updated in the latest commit.

File: monai/data/dataset.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/data/dataset.py]: Cool, thanks for your finding!
I added support to LMDBDataset in the latest commit and unit tests.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review and suggestions.
I updated the PR according to your comments.
Could you please help review it again?
I already added multi-processing test based on workers of DataLoader:
https://github.com/Project-MONAI/MONAI/pull/2436/files#diff-eea046ab7338e5acd9b172b24ec5a704a755dd10bfae2056b046d598d064ecf9R97

Thanks in advance.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

BTW, I already added the doc-string in `CacheDataset` to highlight that it requires `persistent_workers=False` in DataLoader:
https://github.com/Project-MONAI/MONAI/pull/2436/files#diff-e682e654b07b9d753cac73f5a782e70644c82a84d30ebc757ec294a261845108R541

Thanks.

File: monai/data/dataset.py [hunk 4]
```diff
@@ -130,6 +130,9 @@ class PersistentDataset(Dataset):
     Subsequent uses of a dataset directly read pre-processed results from `cache_dir`
     followed by applying the random dependant parts of transform processing.
 
+    If want to update the input data and recompute cache content during training, just call `update_data()`
```
Comment (ericspod) [inline monai/data/dataset.py:133]: Might be easier to say "During training call `update_data()` to update input data and recompute cache content.

File: monai/data/dataset.py [hunk 5]
```diff
@@ -458,6 +479,7 @@ def _cachecheck(self, item_transformed):
         if self._read_env is None:
             self._read_env = self._fill_cache_start_reader()
         with self._read_env.begin(write=False) as txn:
+            print("!!!!!!!!", item_transformed)
```
Comment (ericspod) [inline monai/data/dataset.py:482]: Print statement shouldn't be here?

File: monai/data/dataset.py [hunk 6]
```diff
@@ -515,6 +537,9 @@ class CacheDataset(Dataset):
     ``RandCropByPosNegLabeld`` and ``ToTensord``, as ``RandCropByPosNegLabeld`` is a randomized transform
     and the outcome not cached.
 
+    If want to update the input data and recompute cache content during training, just call `update_data()`
```
Comment (ericspod) [inline monai/data/dataset.py:540]: Same with this sentence and following ones, it can be condensed down a bit.

File: monai/data/dataset.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/data/dataset.py:482]: sorry for the typo, removed it in the latest commit.

Thanks.

File: monai/data/dataset.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/data/dataset.py:133]: sure, updated in the latest commit.

Thanks.

File: monai/data/dataset.py (hunk 6 above)
Comment (Nic-Ma) [reply monai/data/dataset.py]: Sure, updated all related in the latest commit.

Thanks.

**Review by wyli — APPROVED** [review]
thanks, just need to change the docstring `update_data()` to `set_data()`

**Pull request merged**